File: linked_list/linked_list.py
"""
Class that represents a single linked
list node that holds a single value
and a reference to the next node in the list
"""
class Node:
  def __init__(self, value=None, next_node=None):
    self.value = value
    self.next_node = next_node

  def get_value(self):
    return self.value

  def get_next(self):
    return self.next_node

  def set_next(self, new_next):
    self.next_node = new_next

class LinkedList:
  def __init__(self):
    self.head = None
    self.tail = None

  def add_to_tail(self, value):
    node = Node(value)
    if not self.head:
      self.head = node
      self.tail = node
    else:
      self.tail.set_next(node)
      self.tail = node
    

  def remove_head(self):
    curr_node = self.head
    if not self.head:
      return None
    if not self.head.get_next():
      x = self.head
      self.head = None
      self.tail = None
      return x.get_value()
    cur_value = self.head.get_value()
    self.head = self.head.get_next()
    return cur_value
    

  def contains(self, value):
    curr_node = self.head
    while True:
      if not curr_node:
        return False
      elif curr_node.value == value:
        return True
      else:
        curr_node = curr_node.get_next()
    

  def get_max(self):
    if not self.head:
      return None
    curr_node = self.head.get_next()
    max_node = self.head
    while curr_node:
      if curr_node.get_value() > max_node.get_value():
        max_node = curr_node
      else:
        curr_node = curr_node.get_next()

    return max_node.get_value()
    

File: linked_list/linked_list.js
class Node {
  constructor(value=null, nextNode=null) {
    this.value = value;
    this.nextNode = nextNode;
  }

  getValue() {
    return this.value;
  }

  getNext() {
    return this.nextNode;
  }

  setNext(newNext) {
    this.nextNode = newNext
  }
}

class LinkedList {
  constructor(){
    this.head = null;
    this.tail = null;
  }

  addToTail(value) {
    const newNode = new Node(value);
    if(this.tail) {
      this.tail.setNext(newNode);
    }
    if(!this.head) {
      this.head = newNode;
    }
    this.tail = newNode
  }

  removeHead() {
    const oldHead = this.head;
    let returnValue = null;
    if(this.head){
      if(this.head === this.tail) {
        returnValue = oldHead.getValue();
        this.head = null;
        this.tail = null;
      } else {
        const newHead = oldHead.getNext();
        this.head = newHead
        returnValue = oldHead.getValue();
      }
    }
    return returnValue;
  }

  contains(value) {
    let found = false;
    if(this.head) {
      let current = this.head;
      if(this.head === this.tail) {
        if(this.head.getValue() === value) {
          found = true;
        }
      } else {
        while(current && found === false) {
          if(current.getValue() === value){
            found = true;
          }
          current = current.getNext()
        }
      }
    }
    return found
  }

  getMax() {
    let max = null;
    if(this.head) {
      let current = this.head;
      if(this.head === this.tail) {
        max = this.head.getValue()
      } else {
        while(current){
          if(!max || current.getValue() > max) {
            max = current.getValue();
          }
          current = current.getNext();
        }
      }
    }
    return max;
  }
}

const myList = new LinkedList();
//TEST: adding to tail
// myList.addToTail(1);
// console.log('1:', myList.head.getValue());
// console.log('1:', myList.tail.getValue());
// myList.addToTail(2);
// console.log('2:', myList.tail.getValue());
// console.log('1:', myList.head.getValue());

//TEST: contains
// myList.addToTail(1);
// myList.addToTail(2);
// myList.addToTail(5);
// myList.addToTail(10);
// console.log('True:', myList.contains(10));
// console.log('True:', myList.contains(2));
// console.log('False:', myList.contains(1000));

//TEST remove head
// myList.addToTail(10);
// myList.addToTail(20);
// console.log('10:', myList.removeHead());
// console.log('False:', myList.contains(10));
// console.log('20:', myList.removeHead());
// console.log('False:', myList.contains(20));
//
// myList.addToTail(10);
// console.log('10:', myList.removeHead(10));
// console.log('null:', myList.head);
// console.log('null:', myList.tail);
// console.log('null:', myList.removeHead())

//TEST getMax()
// console.log('null:', myList.getMax());
// myList.addToTail(100);
// console.log('100:', myList.getMax());
// myList.addToTail(55);
// console.log('100:', myList.getMax());
// myList.addToTail(101);
// console.log('101:', myList.getMax());
